project_recalculate/models/project_project.py

```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+# For copyright and license notices, see __openerp__.py file in root directory
+##############################################################################
+
+from openerp import models, fields, api, _
+from openerp.exceptions import Warning
+
+
+class ProjectProject(models.Model):
+    _inherit = 'project.project'
+
+    calculation_type = fields.Selection(
+        [('date_begin', 'Date begin'), ('date_end', 'Date end')],
+        default='date_begin', string='Calculation type',
+        help='How to calculate tasks with date start or date end references')
+
+    def _dates_prepare(self):
+        """
+            Prepare project start or end date, looking into tasks list
+            and depending on project calculation_type
+            - if calculation_type == 'date_begin':
+                project end date = latest date from tasks end dates
+            - if calculation_type == 'date_end':
+                project start date = earliest date from tasks start dates
+
+            NOTE: Do not perform any write operations to DB
+        """
+        vals = {}
+        self.ensure_one()
+        # Here we consider all project task, the ones in a stage with
+        # fold = False and the ones with fold = True
+        dec = fields.Datetime.from_string
+        enc = fields.Date.to_string
+        if len(self.tasks) > 0:
```

Comment:
And you don't even need it, saving one indentation level, because you check afterwards if min_date_start or max_date_end